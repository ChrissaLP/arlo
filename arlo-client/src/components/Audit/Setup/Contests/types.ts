export interface IChoiceValues {
  id?: string
  name: string
  numVotes: string | number
}

export interface IContestValues {
  name: string
  isTargeted: boolean
  totalBallotsCast: string
  numWinners: string
  votesAllowed: string
  choices: IChoiceValues[]
}

export interface IValues {
  contests: IContestValues[]
}
