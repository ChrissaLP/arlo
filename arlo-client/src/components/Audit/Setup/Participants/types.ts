export interface IValues {
  state: string
}
