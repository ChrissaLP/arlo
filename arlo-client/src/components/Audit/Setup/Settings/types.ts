import { IAuditSettings } from '../../../../types'

export interface IValues {
  electionName: IAuditSettings['electionName']
  online: IAuditSettings['online']
  randomSeed: IAuditSettings['randomSeed']
  riskLimit: IAuditSettings['riskLimit']
}
