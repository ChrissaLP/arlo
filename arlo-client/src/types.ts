export interface CreateAuditParams {
  electionId: string
}

export interface AuditFlowParams extends CreateAuditParams {
  token: string
  round?: string
  ballot?: string
}

export interface Candidate {
  id: string
  name: string
  numVotes: number | string
}

export interface SampleSizeOption {
  size: number | string
  prob: number | null
  type: string | null
}

export interface Contest {
  id: string
  name: string
  choices: Candidate[]
  totalBallotsCast: string
}

export interface AuditMember {
  name: string
  affiliation: 'DEM' | 'REP' | 'LIB' | 'IND' | ''
}

export interface Ballot {
  tabulator: string
  batch: string
  record: string
  status: 'AUDITED' | 'NOT_AUDITED'
  vote: 'YES' | 'NO' | 'NO_CONSENSUS' | 'NO_VOTE' | null
}

export interface AuditBoard {
  id: string
  name: string
  members: [AuditMember, AuditMember] | []
  ballots?: Ballot[]
}

export interface BallotManifest {
  filename: string | null
  numBallots: number | '' | null
  numBatches: number | '' | null
  uploadedAt: string | null
}

export interface Jurisdiction {
  id: string
  name: string
  contests: string[]
  auditBoards: AuditBoard[]
  ballotManifest?: BallotManifest
}

export interface RoundContest {
  id: string
  results: {
    [key: string]: number
  }
  sampleSizeOptions: SampleSizeOption[] | null
  sampleSize: number | null
  endMeasurements: {
    isComplete: null | boolean
    pvalue: null | number
  }
}

export interface Round {
  name?: string
  randomSeed?: string
  riskLimit?: number
  contests: RoundContest[]
  startedAt: string
  endedAt: string | null
  jurisdictions?: {
    [key: string]: {
      numBallots: number
    }
  }
}

export interface Audit {
  name: string
  riskLimit: string
  randomSeed: string
  contests: Contest[]
  jurisdictions: Jurisdiction[]
  rounds: Round[]
}
