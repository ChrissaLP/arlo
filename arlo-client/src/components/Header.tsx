import React, { useContext } from 'react'
import styled from 'styled-components'
import {
  Navbar,
  NavbarGroup,
  NavbarHeading,
  Alignment,
} from '@blueprintjs/core'
import { Route } from 'react-router-dom'
import { useAuth0, IAuth0Context } from '../react-auth0-spa'
import FormButton from './Form/FormButton'
import UserContext from '../UserContext'

const ButtonBar = styled.div`
  display: inline-block;
  margin-right: 10px;
`

const Nav = styled(Navbar)`
  width: 100%;

  .bp3-navbar-heading img {
    height: 35px;
    padding-top: 8px;
  }
`

const Header: React.FC<{}> = () => {
  const {
    isAuthenticated,
    loginWithRedirect,
    logout,
  } = useAuth0() as IAuth0Context
  const user = useContext(UserContext)
  return (
    <Nav fixedToTop>
      <NavbarGroup align={Alignment.LEFT}>
        <NavbarHeading>
          <img src="/arlo.png" alt="Arlo, by VotingWorks" />
        </NavbarHeading>
      </NavbarGroup>
      {isAuthenticated && (
        <NavbarGroup align={Alignment.LEFT}>
          <NavbarHeading>Welcome, {user.name}</NavbarHeading>
        </NavbarGroup>
      )}
      <NavbarGroup align={Alignment.RIGHT}>
        <ButtonBar id="reset-button-wrapper" />
        <Route
          path="/election"
          render={() =>
            !isAuthenticated && (
              <FormButton size="sm" onClick={() => loginWithRedirect({})}>
                Log in
              </FormButton>
            )
          }
        />
        {isAuthenticated && (
          <FormButton size="sm" onClick={() => logout()}>
            Log out
          </FormButton>
        )}
      </NavbarGroup>
    </Nav>
  )
}

export default Header
