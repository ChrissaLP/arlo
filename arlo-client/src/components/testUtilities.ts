import React from 'react'
import { RenderResult, act, render } from '@testing-library/react'
import { createLocation, createMemoryHistory } from 'history'
import { match as routerMatch } from 'react-router-dom'
import { IAudit, IUserMeta, IRound, IAuditSettings, IBallot } from '../types'
import { IJurisdictions } from './Audit/Setup/useParticipantsApi'
import { IContests } from './Audit/Setup/Contests/types'

type MatchParameter<Params> = { [K in keyof Params]?: string }

const generateUrl = <Params extends MatchParameter<Params>>(
  path: string,
  params: Params
): string => {
  let tempPath = path

  for (const param in params) {
    /* istanbul ignore else */
    if (Object.prototype.hasOwnProperty.call(params, param)) {
      const value = params[param]
      tempPath = tempPath.replace(`:${param}`, value as NonNullable<
        typeof value
      >)
    }
  }

  return tempPath
}

export const routerTestProps = <Params extends MatchParameter<Params> = {}>(
  path: string,
  params: Params
) => {
  const match: routerMatch<Params> = {
    isExact: false,
    path,
    url: generateUrl(path, params),
    params,
  }
  const history = createMemoryHistory()
  const location = createLocation(match.url)

  return { history, location, match }
}

/** Credit to https://stackoverflow.com/a/56452779 for solution to mocking React Router props */

export const regexpEscape = (s: string) => {
  /* eslint-disable no-useless-escape */
  return s.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')
}

export const asyncActRender = async (
  component: React.ReactElement
): Promise<RenderResult> => {
  let result: RenderResult
  await act(async () => {
    result = render(component)
  })
  return result!
}

export const generateApiMock = ({
  statusReturn,
  authReturn,
  roundReturn,
  jurisdictionReturn,
  settingsReturn,
  ballotsReturn,
  contestsReturn,
}: {
  statusReturn?: IAudit | Error | { status: 'ok' }
  authReturn?: IUserMeta | Error
  roundReturn?: { rounds: IRound[] } | Error | { status: 'ok' }
  jurisdictionReturn?:
    | { jurisdictions: IJurisdictions }
    | Error
    | { status: 'ok' }
  settingsReturn?: IAuditSettings | Error
  ballotsReturn?: { ballots: IBallot[] } | Error
  contestsReturn?: IContests | Error
}) => async (
  endpoint: string
): Promise<
  | IAudit
  | IUserMeta
  | { rounds: IRound[] }
  | { jurisdictions: IJurisdictions }
  | IAuditSettings
  | { ballots: IBallot[] }
  | IContests
  | Error
  | { status: 'ok' }
> => {
  if (endpoint === '/election/1/audit/status' && statusReturn)
    return statusReturn
  if (endpoint === '/auth/me' && authReturn) return authReturn
  if (endpoint === '/election/1/round' && roundReturn) return roundReturn
  if (endpoint === '/election/1/jurisdiction' && jurisdictionReturn)
    return jurisdictionReturn
  if (endpoint === '/election/1/settings' && settingsReturn)
    return settingsReturn
  if (
    endpoint.match(
      /\/election\/[^/]+\/jurisdiction\/[^/]+\/round\/[^/]+\/ballot-list/
    ) &&
    ballotsReturn
  )
    return ballotsReturn
  if (endpoint === '/election/1/contest' && contestsReturn)
    return contestsReturn
  return new Error(`missing mock for ${endpoint}`)
}

export default {
  routerTestProps,
}
