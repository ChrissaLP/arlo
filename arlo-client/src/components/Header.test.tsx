import React from 'react'
import { render } from '@testing-library/react'
import { RouteComponentProps } from 'react-router-dom'
import Header from './Header'
import { ICreateAuditParams } from '../types'
import { routerTestProps } from './testUtilities'

const routeProps: RouteComponentProps<ICreateAuditParams> = routerTestProps(
  '/election/:electionId',
  {
    electionId: '1',
  }
)

it('renders correctly', () => {
  const { container } = render(<Header {...routeProps} />)
  expect(container).toMatchSnapshot()
})
