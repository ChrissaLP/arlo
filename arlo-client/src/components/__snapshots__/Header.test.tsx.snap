// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<div>
  <div
    class="bp3-navbar bp3-fixed-top sc-htpNat ocefa"
  >
    <div
      class="bp3-navbar-group bp3-align-left"
    >
      <div
        class="bp3-navbar-heading"
      >
        <a
          href="/"
        >
          <img
            alt="Arlo, by VotingWorks"
            src="/arlo.png"
          />
        </a>
      </div>
    </div>
    <div
      class="bp3-navbar-group bp3-align-right"
    >
      <div
        class="sc-bwzfXH kwlduM"
        id="reset-button-wrapper"
      />
    </div>
  </div>
</div>
`;
